File: core/src/main/java/dev/vibris/core/RuntimeAwaiter.java
package dev.vibris.core;

import dev.vibris.api.CaptureResult;
import dev.vibris.protocol.v1.ErrorCode;
import dev.vibris.protocol.v1.JobResult;
import dev.vibris.protocol.v1.JobTimings;

import java.util.concurrent.CancellationException;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

final class RuntimeAwaiter {
    private final CoreProbe probe;

    RuntimeAwaiter(CoreProbe probe) {
        this.probe = probe;
    }

    CaptureResult capture(CompletionStage<CaptureResult> stage, CoreJob job, long deadline)
        throws RuntimeJobExecutor.Failure {
        return await(stage, job, deadline, ErrorCode.CAPTURE_FAILED);
    }

    <T> T await(CompletionStage<T> stage, CoreJob job, long deadline) throws RuntimeJobExecutor.Failure {
        return await(stage, job, deadline, ErrorCode.INTERNAL_ERROR);
    }

    JobResult withTimings(CoreJob job, JobResult result, long startedAtUnixMs, long startedNanos) {
        long completedAtUnixMs = System.currentTimeMillis();
        long executionMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startedNanos);
        long queueMs = TimeUnit.NANOSECONDS.toMillis(startedNanos - job.acceptedNanos);
        return result.toBuilder().setTimings(JobTimings.newBuilder()
            .setStartedAtUnixMs(startedAtUnixMs)
            .setCompletedAtUnixMs(completedAtUnixMs)
            .setQueueMs(queueMs)
            .setExecutionMs(executionMs)
            .setTotalMs(Math.addExact(queueMs, executionMs))).build();
    }

    private <T> T await(CompletionStage<T> stage, CoreJob job, long deadline, ErrorCode operationFailure)
        throws RuntimeJobExecutor.Failure {
        CompletableFuture<T> future = stage.toCompletableFuture();
        try {
            if (job.cancellation.token().isCancellationRequested()) throw new CancellationException();
            long remaining = deadline == Long.MAX_VALUE ? Long.MAX_VALUE : deadline - System.nanoTime();
            if (remaining <= 0) throw new TimeoutException();
            return remaining == Long.MAX_VALUE ? future.get() : future.get(remaining, TimeUnit.NANOSECONDS);
        } catch (TimeoutException exception) {
            job.cancellation.cancel();
            awaitCancellation(future);
            probe.event(job.requestId, "SAFE_POINT_TIMEOUT");
            throw new RuntimeJobExecutor.Failure(ErrorCode.EXECUTION_TIMEOUT, "Job execution timed out.");
        } catch (InterruptedException exception) {
            Thread.currentThread().interrupt();
            job.cancellation.cancel();
            throw new RuntimeJobExecutor.Failure(ErrorCode.CANCELLED, "Job execution was interrupted.");
        } catch (CancellationException exception) {
            throw new RuntimeJobExecutor.Failure(ErrorCode.CANCELLED, "Job execution was cancelled.");
        } catch (ExecutionException | CompletionException exception) {
            Throwable cause = exception.getCause();
            if (cause instanceof CancellationException || job.cancellation.token().isCancellationRequested()) {
                throw new RuntimeJobExecutor.Failure(ErrorCode.CANCELLED, "Job execution was cancelled.");
            }
            if (operationFailure == ErrorCode.CAPTURE_FAILED) throw CaptureJobExecutor.failure(cause);
            throw new RuntimeJobExecutor.Failure(operationFailure, "Runtime adapter operation failed.");
        }
    }

    private static void awaitCancellation(CompletableFuture<?> future) throws RuntimeJobExecutor.Failure {
        try {
            future.get(5, TimeUnit.SECONDS);
        } catch (CancellationException | ExecutionException ignored) {
        } catch (InterruptedException exception) {
            Thread.currentThread().interrupt();
            throw new RuntimeJobExecutor.Failure(
                ErrorCode.CANCELLED, "Interrupted while waiting for a runtime safe point.");
        } catch (TimeoutException exception) {
            throw new RuntimeJobExecutor.Failure(
                ErrorCode.INTERNAL_ERROR, "Runtime did not stop at a cancellation safe point.");
        }
    }
}

File: core/src/main/kotlin/dev/vibris/core/RuntimeAwaiter.kt
package dev.vibris.core

import dev.vibris.api.CaptureResult
import dev.vibris.protocol.v2.ErrorCode
import dev.vibris.protocol.v2.JobResult
import dev.vibris.protocol.v2.JobTimings
import java.util.concurrent.CancellationException
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import java.util.concurrent.CompletionStage
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

internal class RuntimeAwaiter(private val probe: CoreProbe) {
    @Throws(RuntimeJobExecutor.Failure::class)
    fun capture(stage: CompletionStage<CaptureResult>, job: CoreJob, deadline: Long): CaptureResult =
        await(stage, job, deadline, ErrorCode.ERROR_CODE_CAPTURE_FAILED)

    @Throws(RuntimeJobExecutor.Failure::class)
    fun <T> await(stage: CompletionStage<T>, job: CoreJob, deadline: Long): T =
        await(stage, job, deadline, ErrorCode.ERROR_CODE_INTERNAL)

    fun withTimings(job: CoreJob, result: JobResult, startedAtUnixMs: Long, startedNanos: Long): JobResult {
        val completedAtUnixMs = System.currentTimeMillis()
        val executionMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startedNanos)
        val queueMs = TimeUnit.NANOSECONDS.toMillis(startedNanos - job.acceptedNanos)
        return result.toBuilder()
            .setTimings(
                JobTimings.newBuilder()
                    .setAcceptedAtUnixMs(job.acceptedAtUnixMs)
                    .setStartedAtUnixMs(startedAtUnixMs)
                    .setCompletedAtUnixMs(completedAtUnixMs)
                    .setQueueMs(queueMs)
                    .setExecutionMs(executionMs)
                    .setTotalMs(Math.addExact(queueMs, executionMs)),
            )
            .build()
    }

    @Throws(RuntimeJobExecutor.Failure::class)
    private fun <T> await(
        stage: CompletionStage<T>,
        job: CoreJob,
        deadline: Long,
        operationFailureCode: ErrorCode,
    ): T {
        val future = stage.toCompletableFuture()
        try {
            if (job.cancellation.token().isCancellationRequested()) {
                throw CancellationException()
            }
            val remaining = if (deadline == Long.MAX_VALUE) Long.MAX_VALUE else deadline - System.nanoTime()
            if (remaining <= 0) {
                throw TimeoutException()
            }
            return if (remaining == Long.MAX_VALUE) future.get() else future.get(remaining, TimeUnit.NANOSECONDS)
        } catch (_: TimeoutException) {
            job.cancellation.cancel()
            awaitCancellation(future)
            probe.event(job.requestId, "SAFE_POINT_TIMEOUT")
            throw RuntimeJobExecutor.Failure(ErrorCode.ERROR_CODE_EXECUTION_TIMEOUT, "Job execution timed out.")
        } catch (_: InterruptedException) {
            job.cancellation.cancel()
            awaitCancellation(future)
            Thread.currentThread().interrupt()
            throw RuntimeJobExecutor.Failure(ErrorCode.ERROR_CODE_CANCELLED, "Job execution was interrupted.")
        } catch (_: CancellationException) {
            throw RuntimeJobExecutor.Failure(ErrorCode.ERROR_CODE_CANCELLED, "Job execution was cancelled.")
        } catch (exception: ExecutionException) {
            throw operationFailure(exception.cause, job, operationFailureCode)
        } catch (exception: CompletionException) {
            throw operationFailure(exception.cause, job, operationFailureCode)
        }
    }

    private fun operationFailure(
        cause: Throwable?,
        job: CoreJob,
        code: ErrorCode,
    ): RuntimeJobExecutor.Failure {
        if (cause is CancellationException || job.cancellation.token().isCancellationRequested()) {
            return RuntimeJobExecutor.Failure(ErrorCode.ERROR_CODE_CANCELLED, "Job execution was cancelled.")
        }
        return if (code == ErrorCode.ERROR_CODE_CAPTURE_FAILED) {
            CaptureJobExecutor.failure(cause)
        } else {
            RuntimeJobExecutor.Failure(code, "Runtime adapter operation failed.")
        }
    }

    @Throws(RuntimeJobExecutor.Failure::class)
    private fun awaitCancellation(future: CompletableFuture<*>) {
        try {
            future.join()
        } catch (_: CancellationException) {
        } catch (_: CompletionException) {
        }
    }
}
